File: source/dotnet/Library/AdaptiveCards/AdaptiveImageSet.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
using System.Collections.Generic;
using System.ComponentModel;
using System.Xml.Serialization;
using Newtonsoft.Json;

namespace AdaptiveCards
{
    /// <summary>
    /// Represents an ImageSet element.
    /// </summary>
    [XmlType(TypeName = AdaptiveImageSet.TypeName)]
    public class AdaptiveImageSet : AdaptiveElement
    {
        /// <inheritdoc />
        public const string TypeName = "ImageSet";

        /// <inheritdoc />
        [XmlIgnore]
        public override string Type { get; set; } = TypeName;

        /// <summary>
        /// Collection of images to display.
        /// </summary>
        [JsonRequired]
        [XmlElement(typeof(AdaptiveImage), ElementName = AdaptiveImage.TypeName)]
        public List<AdaptiveImage> Images { get; set; } = new List<AdaptiveImage>();

        /// <summary>
        /// Specifies the <see cref="AdaptiveImageSize"/> of each image in the set.
        /// </summary>
        [JsonProperty(DefaultValueHandling = DefaultValueHandling.IgnoreAndPopulate)]
        [XmlAttribute]
        [DefaultValue(typeof(AdaptiveImageSize), "auto")]
        public AdaptiveImageSize ImageSize { get; set; }
    }
}

File: source/uwp/AdaptiveCardsObjectModel/lib/AdaptiveImageSet.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
#include "pch.h"

#include "AdaptiveImageSet.h"
#include "AdaptiveImageSet.g.cpp"
#include "AdaptiveImage.h"

namespace winrt::AdaptiveCards::ObjectModel::Uwp::implementation
{
    AdaptiveImageSet::AdaptiveImageSet(const std::shared_ptr<::AdaptiveCards::ImageSet>& sharedImageSet)
    {
        Images = GenerateVectorProjection<implementation::AdaptiveImage>(sharedImageSet->GetImages());
        ImageSize = static_cast<Uwp::ImageSize>(sharedImageSet->GetImageSize());
        InitializeBaseElement(sharedImageSet);
    }

    std::shared_ptr<::AdaptiveCards::BaseCardElement> AdaptiveImageSet::GetSharedModel()
    {
        auto imageSet = std::make_shared<::AdaptiveCards::ImageSet>();
        CopySharedElementProperties(*imageSet);
        imageSet->SetImageSize(static_cast<::AdaptiveCards::ImageSize>(ImageSize.get()));
        imageSet->GetImages() = GenerateSharedVector<implementation::AdaptiveImage, ::AdaptiveCards::Image>(Images.get());
        return imageSet;
    }
}
